### ISSUE TITLE
Publish Kafka message to the right topic

### ISSUE THREAD

**Issue by Gnucki** [php-enqueue/messenger-adapter#9]
I publish a Kafka message like this one:
```php
$this->bus->dispatch(new \Enqueue\RdKafka\RdKafkaMessage('foo', ['topic' => 'test']));
```
I expect this to publish my message `foo` in topic `test`.

In fact, in `send` method of `Enqueue\MessengerAdapter\QueueInteropTransport` my message (wrapped in an `Symfony\Component\Messenger\Envelope`) is encoded by an instance of `Symfony\Component\Messenger\Transport\Serialization\Serializer` which gives:
```php
array(2) {
  ["body"]=>
  string(72) ""{\"body\":\"plop\",\"properties\":{\"topic\":\"test\"},\"headers\":[]}""
  ["headers"]=>
  array(1) {
    ["type"]=>
    string(30) "Enqueue\RdKafka\RdKafkaMessage"
  }
}
```

After creation of a new `\Enqueue\RdKafka\RdKafkaMessage` (in the same function), I finally have the following sent message:
```php
object(Enqueue\RdKafka\RdKafkaMessage)#402 (7) {
  ["body":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  string(72) ""{\"body\":\"plop\",\"properties\":{\"topic\":\"test\"},\"headers\":[]}""
  ["properties":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  array(0) {
  }
  ["headers":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  array(1) {
    ["type"]=>
    string(30) "Enqueue\RdKafka\RdKafkaMessage"
  }
  ["redelivered":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  bool(false)
  ["partition":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  NULL
  ["key":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  NULL
  ["kafkaMessage":"Enqueue\RdKafka\RdKafkaMessage":private]=>
  NULL
}
```
My original message has been serialized in the body of the sent message and the target topic is the default one (i.e. `messages`) instead of `test`.

I see that the destination topic comes from `$this->options` but I do not know how to set that options.

It is difficult for me to know the best way (implement a new encoder, define some options, upgrade code, ...) to achieve what I want to do. That's why I'm asking you for help.

**Comment by sroze** [comment]
The message you dispatch in the messenger bus should not be an instance of the `Enqueue\RdKafka\RdKafkaMessage`. It should be your own domain object.

> I see that the destination topic comes from $this->options but I do not know how to set that options.

These options are coming from the DSN of the Messenger transport. To change the default topic, you can do:
```yaml
# config/packages/messenger.yaml
framework:
    messenger:
        transports:
            amqp: enqueue://default?topic[name]=test
```

Though, if you want to change the destination of a specific message (while having a different default topic) it's trickier. The easiest being to create another Messenger transport for it. Otherwise, the `RdKafkaMessage` will be constructed by Enqueue's transport. As you mentioned, the `properties` and `headers` are coming from the encoder directly and the default one does not return any property. I see a few different options in here:

We could leverage [the `Envelope` mechanism of the Messenger component](https://github.com/symfony/symfony-docs/pull/9757). We'd have to update the producer to support them and you would do something like this when dispatching the message:
```php
$bus->dispatch((new Envelope($yourMessageObject))->with(new EnqueueProperties(['topic' => 'test'])));
```

**Comment by Gnucki** [comment]
I made a trivial implementation around your idea (without unit testing for the moment).
Tell me if the base is OK and what I can do to make it more "industrial" if this is the case!

Use:
```php
use Symfony\Component\Messenger\Envelope;
use Enqueue\MessengerAdapter\EnvelopeItem\Topic;
use App\Message\TestMessage;

$message = new TestMessage('foo');
$this->bus->dispatch((new Envelope($message))->with(new Topic('test')));
```

Config:
```yaml
# messenger.yaml
framework:
    messenger:
        transports:
            kafka_test: enqueue://rdkafka?queue[name]=test
        routing:
            'App\Message\TestMessage': kafka_test
```
```yaml
# enqueue.yaml
enqueue:
    async_events:
        enabled: true
    transport:
        default: '%env(ENQUEUE_DSN)%'
        rdkafka:
            global:
                group.id: 123
                metadata.broker.list: 'localhost:9092'
            topic:
                auto.offset.reset: beginning
    client: ~
```

**Comment by Gnucki** [comment]
#12

**Comment by sroze** [comment]
Fixed by #12. Thanks again :)
